### PR TITLE
fixes & improvements

### PR DESCRIPTION
**Pull request by ideas-into-software**
Hi, 

I just committed few fixes and improvements which I needed to make. Those are: 

1. FIX: file not being closed 
In `lib/jekyll-linkpreview.rb`, function `save_cache_file` had a problem due to file not being closed. This error could be seen on first generation / when no cache entry existed for given URL and attempt to create then read cache entry for given URL was made.

2. FIX: variable not being extracted properly 
This error prevented me from passing URLs as variables to the linkpreview tag. Now either string or a variable can be passed with no problems.

3. IMPROVEMENT: ability to customize layout via template - instead of using those embedded in plugin source code
More information about this in code and the updated README.md file.

4. Other: 
* updated 'spec/jekyll-linkpreview_spec.rb'
* bumped 'lib/jekyll-linkpreview/version.rb' to version 0.3.0
* bumped 'jekyll-linkpreview.gemspec' jekyll version to '3.8.6'
* updated 'README.md'

Some of these fixes and improvements actually coinicide with issues I noticed you had open, which you can now close, i.e.: 
 * https://github.com/ysk24ok/jekyll-linkpreview/issues/2
 * https://github.com/ysk24ok/jekyll-linkpreview/issues/3

Regards,
Michael

### CODE REVIEW

**Comment by ysk24ok** [comment]
> FIX: variable not being extracted properly
> This error prevented me from passing URLs as variables to the linkpreview tag.

Would you please add that case to tests ? Because it seems existing test cases passing a variable work properly.

**Comment by ideas-into-software** [comment]
Hi,

I did update test case you've had ('spec/jekyll-linkpreview_spec.rb') to reflect changes I made.

Unfortunately, I do not have time to create new test cases now - however, if you wish to verify, it's enough to create new Jekyll site (e.g. jekyll new jekyll-linkpreview-testsite), add the plugin (currently released version, i.e. 0.2.0), then attempt to use variables instead of strings, and you'll see how this works, i.e. it does not. Also, you have an existing open issue https://github.com/ysk24ok/jekyll-linkpreview/issues/2 which talks about the same thing, i.e. not being able to use variables in Jekyll sites when using current version of your plugin.

So, to summarize, having this work in an actual deployment (i.e. Jekyll site) is more important than automated test cases which obviously are missing something because the two are not in sync.

Hope this helps.

Michael

**Comment by ysk24ok** [comment]
It seems jekyll-linkpreview v0.2.0 does work with variables.
I tried creating new site by `jekyll new`, installing v0.2.0 and adding following code to index.md,

```
{% assign url = 'https://github.com' %}
{% linkpreview url %}
```

and confirmed it works fine like this.

スクリーンショット 2019-12-13 23 27 19

(the image is so big because no css is applied ...)


I want to know why your attempt to use variables fails. That's why I want you to add a testcase, which fails with v0.2.0 and passes with this pull request.
Could you explain your attempt more precisely?

File: README.md [hunk 1]
```diff
@@ -61,6 +61,32 @@ See https://jekyllrb.com/docs/plugins/installation/ .
...
+  * **link_domain** i.e. `{{ link_domain }}`
+
+### No Open Graph Protocol preview template
+
+Template used for generating previews for pages where Open Graph Protocol metadata **does not exist**.
+
+1. Place `linkpreview_nog.html` file inside `_includes/` folder of your Jekyll site (`_includes/linkpreview_nog.html`)
+
+2. Use built-in **link_url** variable to render URL data, i.e. `{{ link_url }}`
+
```
Comment (ysk24ok) [inline README.md:89]: Thank you for the detailed explanation!

File: lib/jekyll-linkpreview.rb [hunk 2]
```diff
@@ -122,7 +96,7 @@ def get_properties(url)
 
       private
       def get_url_from(context)
-        context.scopes[0].key?(@markup) ? context.scopes[0][@markup] : @markup
+        context[@markup]
```
Comment (ysk24ok) [inline lib/jekyll-linkpreview.rb:99]: Becomes much simpler!
Now users have to pass quoted url like `{% linkpreview 'https://github.com' %}`, but it may be acceptable.

File: lib/jekyll-linkpreview.rb [hunk 3]
```diff
@@ -132,7 +106,69 @@ def load_cache_file(filepath)
 
       protected
       def save_cache_file(filepath, properties)
-        File.open(filepath, 'w').write(JSON.generate(properties))
+        File.open(filepath, 'w') { |f| f.write JSON.generate(properties) }
```
Comment (ysk24ok) [inline lib/jekyll-linkpreview.rb:109]: 👍

File: lib/jekyll-linkpreview.rb [hunk 4]
```diff
@@ -132,7 +106,69 @@ def load_cache_file(filepath)
...
+            <a href="#{url}" target="_blank">#{title}</a>
+          </h2>
+          <div class="jekyll-linkpreview-description">#{description}</div>
+          </div>
+          </div>
+          <div class="jekyll-linkpreview-footer">
+          <a href="//#{domain}" target="_blank">#{domain}</a>
+          </div>
+          </div>
+          </div>
```
Comment (ysk24ok) [inline lib/jekyll-linkpreview.rb:141]: Indentation should be deleted because it doesn't rendered properly.

File: lib/jekyll-linkpreview.rb [hunk 5]
```diff
@@ -132,7 +106,69 @@ def load_cache_file(filepath)
...
+        template_path = get_linkpreview_nog_template()
+        if File.exist?(template_path)
+          template_file = File.read template_path
+          site = context.registers[:site]
+          template_file = (Liquid::Template.parse template_file).render site.site_payload.merge!({"link_url" => url})
+        else
+          html = <<-EOS
+          <div class="jekyll-linkpreview-wrapper">
+          <p><a href="#{url}" target="_blank">#{url}</a></p>
+          </div>
```
Comment (ysk24ok) [inline lib/jekyll-linkpreview.rb:158]: Indentation should be deleted, same reason above.

**Review by ysk24ok — CHANGES_REQUESTED** [review]
Anyway, thank you for this pull request! Custom template for preview is a cool feature.
But there is a bug when no custom template exists. Indentation in heredoc should be deleted, otherwise rendered HTML will be like this.

スクリーンショット 2019-12-20 20 41 28

**Comment by ysk24ok** [comment]
There is a problem (heredoc indent) in this PR, but I'll fix it after merge.
Anyway, thank you for this PR!

**Comment by ideas-into-software** [comment]
Hi, Sorry I could not help with formatting, I did not have the time to check GitHub since then. Thank you for the merge - glad I could help.
